docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,115 @@
+
+# Volume Healer Controller
+
+**Now the obvious question we expect is "what do we achieve with the new controller?"**
+* Achieve a way to monitor the health of volumes.
+* Ceph-CSI has healer component which take cares of restarting the userspace processes when the node plugin is restarted, but in many places it uses Kubenetes specific code, with this controller model we are moving out the Kubernetes specific code out, and help make Ceph-CSI orchestrator independent.
+* We currently do not have a way to act on failed volumes. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it. For example userspace mounters such as NBD has a process running per volume, this controller should ideally help restart such processes in case if they are not-started or crashed.
+* This will be a generic controller design for any mounters added in the future, for userspace mounters it will offer healing in addition, like restarting the respective processes if the volume goes abnormal. Some example userspace mounters would be rbd-nbd, cephfs-fuse, etc.
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller. It will trigger controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
+    * Fetch all the VolumeAttachment` list
+    * Filter the `VolumeAttachment` list through matching driver name, node name and status attached
+    * For each `VolumeAttachment` get the respective PersistentVolume information and check the criteria of PersistentVolume Bound (also we can have configuration options like mounter type)
+    * Send the RPC request for side-car to check if the volume is healthy along with the PersistentVolume details and may attempt to heal the volume if it is found abnormal
+    * Log the final response returned from the side-car
+
+## CSI-Addons side-car
+* When a nodeplugin container is restarted the side-car will update/create(if not present) [CSIAddonsNode](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/api/v1alpha1/csiaddonsnode_types.go) object.
```

Comment:
```suggestion
* When a nodeplugin pod is restarted the side-car will create [CSIAddonsNode](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/api/v1alpha1/csiaddonsnode_types.go) object.
```

When a pod is restarted, it essentially is deleted (which also deletes corresponding CSIAddonsNode object) and
a new pod is created (with new CSIAddonsNode object).

Reply:
Agree